### ISSUE TITLE
Unclear configuration of storysource with Typescript causes TypeErrors

### ISSUE THREAD

**Issue by aloifolia** [storybookjs/storybook#11793]
**Describe the bug**
The README for the storysource addon is confusing if you are not quite familiar with Webpack. You are supposed to change the `main.js` file like this:

```js
module.exports = {
  addons: ['@storybook/addon-storysource'],
};
```

But also, if you want to use Typescript-based stories, you need to change `main.js` like this:

```js
module.exports = function({ config }) { /* ... */ };
```

I have problems with merging these two exports. My current not working solution looks like this:

```js
module.exports = function ({ config }) {
    config.stories = ['../src/**/*.stories.mdx', '../src/**/*.stories.@(js|jsx|ts|tsx)'];
    config.addons = [
        '@storybook/addon-links',
        '@storybook/addon-actions',
        '@storybook/addon-essentials',
        '@storybook/addon-storysource',
        '@storybook/addon-docs',
    ];

    config.module.rules.push({
        test: /\.stories\.tsx?$/,
        loaders: [
            {
                loader: require.resolve('@storybook/source-loader'),
                options: { parser: 'typescript' },
            },
        ],
        enforce: 'pre',
    });

    return config;
};
```

But this does not work since the passed argument does not have a `config` field.

**To Reproduce**
Install the current storybook version (6.0.0.-rc.x) and try to apply the instructions of the storysource README. Use one of the function-based exports and you will see an error message which indicates that `config` is undefined.

**Expected behavior**
Ideally, the README would better explain how the function-based exports are to be merged with the default record based export. 

**Screenshots**
![Screenshot 2020-08-04 at 12 26 50](https://user-images.githubusercontent.com/1295054/89283693-dbff7800-d64d-11ea-8f7c-fcbf97ac4863.png)

**System:**
```
Environment Info:

  System:
    OS: macOS Mojave 10.14.6
    CPU: (8) x64 Intel(R) Core(TM) i5-8259U CPU @ 2.30GHz
  Binaries:
    Node: 13.12.0 - ~/.nvm/versions/node/v13.12.0/bin/node
    Yarn: 1.22.4 - /usr/local/bin/yarn
    npm: 6.14.4 - ~/.nvm/versions/node/v13.12.0/bin/npm
  Browsers:
    Chrome: 84.0.4147.105
    Firefox: 76.0.1
    Safari: 13.1
  npmPackages:
    @storybook/addon-actions: 6.0.0-rc.13 => 6.0.0-rc.13 
    @storybook/addon-docs: ^6.0.0-rc.16 => 6.0.0-rc.16 
    @storybook/addon-essentials: 6.0.0-rc.13 => 6.0.0-rc.13 
    @storybook/addon-links: 6.0.0-rc.13 => 6.0.0-rc.13 
    @storybook/addon-storysource: 6.0.0-rc.16 => 6.0.0-rc.16 
    @storybook/angular: 6.0.0-rc.13 => 6.0.0-rc.13 
```

**Comment by shilman** [comment]
why do you need to do this?

```js
module.exports = function({ config }) { /* ... */ };
```

**Comment by aloifolia** [comment]
I probably do not need this. However, the documentation (addons/storysource/README.md) tells me to do this:

![Screenshot 2020-08-04 at 17 06 21](https://user-images.githubusercontent.com/1295054/89310438-fdc02580-d674-11ea-9641-531a15ec6e52.png)

So, I just switched to this config instead:

```js
module.exports = {
    stories: ['../src/**/*.stories.mdx', '../src/**/*.stories.@(js|jsx|ts|tsx)'],
    addons: [
        '@storybook/addon-links',
        '@storybook/addon-actions',
        '@storybook/addon-essentials',
        '@storybook/addon-storysource',
        '@storybook/addon-docs',
    ],
    module: {
        rules: [
            {
                test: /\.stories\.tsx?$/,
                loaders: [
                    {
                        loader: require.resolve('@storybook/source-loader'),
                        options: { parser: 'typescript' },
                    },
                ],
                enforce: 'pre',
            },
        ],
    },
};
```

Now the error message is not shown anymore. However, the story panel stays empty. Not sure, if this behaviour has already been reported for Angular projects.

![Screenshot 2020-08-04 at 17 10 36](https://user-images.githubusercontent.com/1295054/89310837-88a12000-d675-11ea-8ac9-5c15c8e0e1db.png)

**Comment by shilman** [comment]
Fixed the docs. Can you check the new docs and see if they make sense?

**Comment by aloifolia** [comment]
Unfortunately, I am still struggling 😕 . I see two problems with the documentation:

1. It does not tell me *where* to put the source loader configuration. Is it `.storybook/main.js` or some other file?
2. It does not clarify how this config is to be merged with the standard configuration (if the config should be added to `main.js`)  which looks like this:

```js
module.exports = {
    stories: ['../src/**/*.stories.mdx', '../src/**/*.stories.@(js|jsx|ts|tsx)'],
    addons: [
        '@storybook/addon-links',
        '@storybook/addon-actions',
        '@storybook/addon-essentials',
        '@storybook/addon-storysource',
        '@storybook/addon-docs',
    ],
};
```

Thus, my third approach, after reading the [custom webpack config guide](https://storybook.js.org/docs/configurations/custom-webpack-config/), was this `main.js` file:

```js
module.exports = {
    stories: ['../src/**/*.stories.mdx', '../src/**/*.stories.@(js|jsx|ts|tsx)'],
    addons: [
        '@storybook/addon-links',
        '@storybook/addon-actions',
        '@storybook/addon-essentials',
        '@storybook/addon-storysource',
        '@storybook/addon-docs',
    ],
    webpackFinal: async (config, { configType }) => {
        config.module.rules.push({
            test: /\.stories\.tsx?$/,
            loaders: [
                {
                    loader: require.resolve('@storybook/source-loader'),
                    options: { parser: 'typescript' },
                },
            ],
            enforce: 'pre',
        });

        return config;
    },
};
```

Since the storysource panel stays blank, I am not sure whether this last approach was theoretically the right call 🤷 .

**Comment by aloifolia** [comment]
So, I just checked the angular-cli example project in the storybook repo. Interestingly enough, it does not contain any source loader customisation for Typescript stories. I am wondering how this works 🤔 .

Anyway, I found `marko-cli/.storybook/main.js` which uses exactly the config merge approach that I described in the last message. So I guess that my last guess *should* be right, although it does not successfully show any source code in the storysource panel.
